File: Scala/algorithmic_toolbox/src/main/scala-2.12/divideandconquer/QuickSort.scala
package divideandconquer

import scala.annotation.tailrec
import scala.util.Random

object QuickSort {

  class Sorting(val array: Array[Int], n: Option[Int]) {
    type Index = Int

    private val length: Int = n.getOrElse(array.length)
    private val rng = new Random(2112L)

    private[this] def swapElems(ix: Index, elem: Int, jy: Index): Unit = {
      array(ix) = array(jy)
      array(jy) = elem
    }

    private[this] def threeWayPartitioning(pivot: Int, start: Index, end: Index): (Index, Index) = {
      @tailrec
      def loop(ix: Index, middleStart: Index, middleEnd: Int): (Index, Index) = {
        if (ix > middleEnd) (middleStart, middleEnd)
        else {
          val elem: Int = array(ix)
          if (elem < pivot) {
            if (ix != middleStart) swapElems(ix, elem, middleStart)
            loop(ix + 1, middleStart + 1, middleEnd)
          } else if (elem > pivot) {
            swapElems(ix, elem, middleEnd)
            loop(ix, middleStart, middleEnd - 1)
          } else loop(ix + 1, middleStart, middleEnd)
        }
      }
      loop(ix = start, middleStart = start, middleEnd = end)
    }

    def quickSort(): Unit = {
      @tailrec
      def loop(stack: List[(Index, Index)]): Unit = {
        if (stack.nonEmpty) {
          val (leftEnd, rightEnd): (Index, Index) = stack.head
          if (leftEnd >= rightEnd) loop(stack.tail)
          else {
            val randomIx: Index = leftEnd + rng.nextInt(rightEnd - leftEnd + 1)
            val pivot: Int = array(randomIx)
            val (middleStart, middleEnd): (Index, Index) = threeWayPartitioning(pivot, leftEnd, rightEnd)
            loop((middleEnd + 1, rightEnd) :: (leftEnd, middleStart - 1) :: stack.tail)
          }
        }
      }
      loop(stack = List((0, length - 1)))
    }
  }

  private def convertToIntArray(line: String): Array[Int] = line.split(" ").map(_.toInt)

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val n: Int = reader.next().toInt
    val lst: Array[Int] = convertToIntArray(reader.next())
    val sorting = new Sorting(lst, Some(n))
    sorting.quickSort()
    println(sorting.array.mkString(" "))
  }
}

File: OCaml/algorithmic_toolbox/divideandconquer/quicksort.ml
let () = Random.init 2112

let convert_to_intlist line = List.map int_of_string (Str.split (Str.regexp " ") line)


let three_way_partitioning arr pivot start_index end_index =
    let swap_elems ix elem jy =
        (arr.(ix) <- arr.(jy);
         arr.(jy) <- elem) in
    let rec loop ix middle_start middle_end =
        if ix > middle_end then (middle_start, middle_end)
        else
            let elem = arr.(ix) in
            if elem < pivot then
                (if ix <> middle_start then swap_elems ix elem middle_start;
                loop (ix + 1) (middle_start + 1) middle_end)
            else if elem > pivot then
                (swap_elems ix elem middle_end;
                loop ix middle_start (middle_end - 1))
            else loop (ix + 1) middle_start middle_end
    in loop start_index start_index end_index


let quicksort arr n =
    let rec loop stack =
        if (stack <> []) then
            let left_end, right_end = List.hd stack in
            if left_end >= right_end then loop (List.tl stack)
            else
                let random_ix = left_end + Random.int (right_end - left_end + 1) in
                let pivot = arr.(random_ix) in
                let middle_start, middle_end = three_way_partitioning arr pivot left_end right_end in
                loop ((middle_end + 1, right_end) :: ((left_end, middle_start - 1) :: (List.tl stack)))
    in loop [(0, n - 1)]


let () =
    let n = read_int() in
    let arr = convert_to_intlist (read_line()) |> Array.of_list in
    quicksort arr n;
    Array.iter (fun elem -> print_int elem; print_string " ") arr;
    print_newline()
